File: python/solvers/dg/initial_guess.py
from itertools import product

from numpy import array, dot, zeros
from scipy.optimize import newton_krylov

from etc.grids import flat_index
from solvers.basis import DERVALS, GAPS
from system import source, system_matrix


def standard_initial_guess(obj, w, *args):
    """ Returns a Galerkin intial guess consisting of the value of q at t=0
    """
    ret = array([w for i in range(obj.N)])
    return ret.reshape(obj.NT, obj.NV)


def stiff_initial_guess(obj, w, dt, dX):
    """ Returns an initial guess based on the underlying equations
    """
    q = zeros([obj.N] * (obj.NDIM + 1) + [obj.NV])
    qt = w.reshape([obj.N] * obj.NDIM + [obj.NV])
    coordList = [range(obj.N)] * obj.NDIM

    for t in range(obj.N):

        dt_ = dt * GAPS[t]

        # loop over the coordinates of each spatial node
        for coords in product(*coordList):

            q_ = qt[coords]     # the value of q at the current spatial node

            # qi[d] holds the coefficients at the nodes lying in a strip in the
            # dth direction, at the current spatial node
            qi = []
            for d in range(obj.NDIM):
                i = flat_index(coords[:d])
                j = flat_index(coords[d + 1:])
                qi.append(qt.reshape(obj.N**d, obj.N, obj.N**(obj.NDIM - d - 1), obj.NV)[i, :, j])

            Mdqdx = zeros(obj.NV)
            for d in range(obj.NDIM):
                dqdxi = dot(DERVALS[coords[d]], qi[d])
                Mdqdx += dot(system_matrix(q_, d, obj.model_params), dqdxi) / dX[d]

            S0 = source(q_, obj.model_params)

            if obj.nk_ig:

                def f(X):

                    S = (S0 + source(X, obj.model_params)) / 2
                    return X - q_ + dt_ * (Mdqdx - S)

                q[(t,) + coords] = newton_krylov(f, q_, f_tol=obj.tol)

            else:
                q[(t,) + coords] = q_ - dt_ * (Mdqdx - S0)

        qt = q[t]

    return q.reshape(obj.NT, obj.NV)

File: cpp/src/solvers/dg/initial_guess.cpp
#include "../../etc/types.h"
#include "../../system/objects/gpr_objects.h"

void standard_initial_guess1(Matr q, Matr w) {
  // Returns a Galerkin intial guess consisting of the value of q at t=0
  for (int i = 0; i < N; i++)
    q.block<N, V>(i * N, 0) = w;
}
void standard_initial_guess2(Matr q, Matr w) {
  // Returns a Galerkin intial guess consisting of the value of q at t=0
  for (int i = 0; i < N; i++)
    q.block<N * N, V>(i * N * N, 0) = w;
}

void stiff_initial_guess(Matr q, Matr w, int NT, double dt, Par &MP) {
  // Returns the initial guess found in DOI: 10.1007/s10915-010-9426-6

  standard_initial_guess1(q, w);
  /*
  Mat qt = w;

  for (int t=0; t<N; t++)
  {
      double DT;
      if (t==0)
          Dt = dt * NODES(0);
      else
          DT = dt * (NODES(t) - NODES(t-1));

      Mat dqdxt = derivs.dot(qt);

      for (int i=0; i<N; i++)
      {
          M = system_conserved(qi, 0, PAR, SYS).dot(dqdxt[i]);
          S = source(qt[i], PAR, SYS);

          if (superStiff)
          {
              f = lambda X: X - qt[i] + dt/dx * M - dt/2 *
  (S+source(X,PAR,SYS));
              q[t,i] = newton_krylov(f, qi, f_tol=TOL);
          }
          else
          {
              q[t,i] = qi - dt/dx * M + dt * Sj;
          }
      }
      qt = q[t];
  }
  */
}
